### PR TITLE
ci: switch from kubeval to kubeconform

### LINKED ISSUE #360: Consider using alternative source for kubeval schemas
[instrumenta/kubernetes-json-schema](https://github.com/instrumenta/kubernetes-json-schema) is noticeably out of date at this point, with the last commit to master landing in April of 2020, this puts the latest provided Kubernetes version at 1.18.1.

As an alternative there is [yannh/kubernetes-json-schema](https://github.com/yannh/kubernetes-json-schema) which is self-updating and currently up-to-date.

It's a trivial change to update the schema source (see: [here](https://github.com/timberio/helm-charts/blob/master/.github/kubeval.sh#L6)).

This would allow for the Helm charts to be evaluated against more, and more recent versions of Kubernetes.

### LINKED ISSUE #1312: Duplicate `nodeSelector` values in datadog chart with CI defaults
**Describe what happened:**
While working on #1199, I noticed that if `kubeconform` is run with `-strict`, errors like this are observed
```
Deployment failed validation: release-name-datadog-cluster-agent error unmarshalling resource: error converting YAML to JSON: yaml: unmarshal errors:
  line 242: key "kubernetes.io/os" already set in map
```

Essentially, I think when `nodeSelector` is specified in a values file, it will be included in addition to the baked in defaults, rather than instead of it. I'm not sure if that's the intended behavior or not.
If it is, maybe the simple fix is to use a different example from the os default for `nodeSelector` in the ci values file, but if one is supposed to replace the other, the fix will be more complicated.

**Describe what you expected:**
No error to be reported

**Steps to reproduce the issue:**

1) Run:
```
% helm dep up charts/datadog && helm template --kube-version "1.22.7" --values charts/datadog/ci/kubeconform-values.yaml charts/datadog
```
2) Run kubeconform on the resulting output with the strict flag.

**Additional environment details (Operating System, Cloud provider, etc):**
Note: I am guessing this is because of stuff like 

https://github.com/DataDog/helm-charts/blob/0e6401a01c732fa303f212f208f3a9289da26ffb/charts/datadog/templates/cluster-agent-deployment.yaml#L483-L487

### PR DESCRIPTION
**Pull request by wyardley**
#### What this PR does / why we need it:
- Switch from kubeval, which is unmaintained and outdated, to kubeconform
- Update kubernetes versions in test matrices
- Use upstream schema for CRDs, so we don't have to ignore missing schemas in validation
- Switch away from deprecated output format
  https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/
- Switch CI nodeSelector examples to use disktype: ssd to avoid creating duplicates with baked in os defaults

Screenshot from local example:
image

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #360
  - fixes #1312

#### Special notes for your reviewer:
Note: I don't know if any chart changes are triggered. I tested this locally, but may be good to do some additional testing in CI if you'd like.

CI changes only; let me know if you want me to bump version / changelog, though I imagine not relevant here.

I'm happy to remove `v1.18.20` or add / remove other versions from the matrices - just let me know.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Chart Version bumped
- [x] `CHANGELOG.md` has been updated

### CODE REVIEW

**Comment by wyardley** [comment]
Note: I just updated it to resolve a deprecation warning with `set-output` - once the job is approved, I will double-check that it looks correct

And, as mentioned before, I can bump the chart version if / when you all want, which is one of the test failures.

I don't think the go test failure is related to my changes?

**Comment by wyardley** [comment]
There is a separate (and similar) matrix of Kube versions for the go integration tests, which I'm happy to edit there / make standard with he ones in this PR if it helps (though I haven't yet)

**Comment by wyardley** [comment]
@clamoriniere would you be able to get someone to review / comment on this, and / or is there anything you'd like me to do / update?

File: .github/kubeconform.sh
Comment (loutPhilipps) [inline .github/kubeconform.sh:34]:
```suggestion
        -output pretty -strict -verbose -kubernetes-version "${KUBERNETES_VERSION#v}" -
```
`-strict` was enabled with `kubeval`, any reason not to enable it with `kubeconform`?

File: .github/workflows/ci.yaml
Comment (loutPhilipps) [inline .github/workflows/ci.yaml:87]: For this I am not sure what the implications are for the agent - I'll let another reviewer comment for these

**Review by loutPhilipps — COMMENTED** [review]
It looks like there's an issue with the output format:
```
Error: Unable to process file command 'output' successfully.
Error: Invalid format 'charts/datadog-operator'
```

File: .github/kubeconform.sh
Comment (wyardley) [reply .github/kubeconform.sh:34]: it has been a while since I put this together / tested it, but it looks like this is probably why:
```
✖ stdin: DaemonSet failed validation: release-name-datadog error unmarshalling resource: error converting YAML to JSON: yaml: unmarshal errors:
  line 879: key "kubernetes.io/os" already set in map
✖ stdin: Deployment failed validation: release-name-datadog-clusterchecks error unmarshalling resource: error converting YAML to JSON: yaml: unmarshal errors:
  line 181: key "kubernetes.io/os" already set in map
✖ stdin: Deployment failed validation: release-name-datadog-cluster-agent error unmarshalling resource: error converting YAML to JSON: yaml: unmarshal errors:
  line 242: key "kubernetes.io/os" already set in map
```

and in fact, this happens a bunch in the configs:
```yaml
      nodeSelector:
        kubernetes.io/os: linux
        kubernetes.io/os: linux
```

My suggestion is that we don't try to fix that in this PR, but I can rebase if someone wants to throw up a fix for it?

File: .github/workflows/ci.yaml
Comment (wyardley) [reply .github/workflows/ci.yaml:87]: This is also pretty old / crusty, so may need to be updated again.

File: .github/kubeconform.sh
Comment (loutPhilipps) [reply .github/kubeconform.sh:34]: I'm fine with that, up to the maintainers of these charts 👍

**Comment by wyardley** [comment]
> It looks like there's an issue with the output format:

Ah! probably shouldn't have tried to fix this error at the same time.
https://stackoverflow.com/questions/74137120/how-to-fix-or-avoid-error-unable-to-process-file-command-output-successfully

I think it should have been `changed` anyway vs `{changed}`; not sure if just changing `{changed}=` to `changed=` will be sufficient, but I can just add in the solution using a heredoc and we can see if that works...

File: .github/workflows/ci.yaml
Comment (wyardley) [inline .github/workflows/ci.yaml:38]: I think this will result in
```
changed
[lines changed]
```
vs `changed=[first line]`. Not 100% sure if that will work or not

File: .github/kubeconform.sh
Comment (wyardley) [reply .github/kubeconform.sh:34]: Are you able to bug someone about it? I can try to take a quick look and see if there's an obvious fix, but it would nice to be able to enable strict here vs. waiting, since it's kind of buried here.

I checked, and it's only `charts/datadog` that has the issue - all the others succeed with `-strict`. If it's a really easy fix, I can try to open up a PR for that

Comment (wyardley) [reply .github/kubeconform.sh:34]:
```
      nodeSelector:
        {{ template "label.os" . }}: {{ .Values.targetSystem }}
      {{- if .Values.agents.nodeSelector }}
{{ toYaml .Values.agents.nodeSelector | indent 8 }}
      {{- end }}
```

Maybe because of this (and similar code in other sections)?

File: .github/workflows/ci.yaml
Comment (wyardley) [reply .github/workflows/ci.yaml:89]: I'm happy to plugin whatever values you all would like here.

I'm not currently using Datadog, but I'm happy to do some small changes if it will help get this shipped.

File: .github/kubeconform.sh
Comment (wyardley) [reply .github/kubeconform.sh:34]: I created #1312 

One possibility is that the chart _intends_ to allow multiple `nodeSelector` statements, and that this isn't intended to override the baked in defaults? It looks like a feature was released in 2.x to always includle the target system nodeSelector; in that case, there's no real bug, and the "fix" would probably be to just use some kind of arbitrary `nodeSelector` value that doesn't match in this PR. I'm happy to make that change if we can confirm this is the intended behavior.

**Comment by wyardley** [comment]
@loutPhilipps if you want to trigger again, we can see if my output fixes now work.

File: .github/kubeconform.sh
Comment (loutPhilipps) [reply .github/kubeconform.sh:34]: Indeed it worked

**Comment by loutPhilipps** [comment]
> Note: I just updated it to resolve a deprecation warning with `set-output` - once the job is approved, I will double-check that it looks correct
> 
> And, as mentioned before, I can bump the chart version if / when you all want, which is one of the test failures.
> 
> I don't think the go test failure is related to my changes?

I believe you do need to bump the chart versions

**Comment by wyardley** [comment]
> I believe you do need to bump the chart versions

I can if you want, but isn't the bump of version just in cases where the chart itself changes? this should (IMO) be a non-triggering change? The actual kubeconform jobs do already run successfully here, so other than the failing CI check, is there a reason the version should be bumped, when the charts themselves are not changing?

The only issue I have with updating the chart versions is that it may make it harder to keep this PR up to date -- if it's ready to ship and they really need to be updated, I can go ahead and do it.

File: .github/workflows/ci.yaml
Comment (wyardley) [reply .github/workflows/ci.yaml:89]: https://github.com/yannh/kubernetes-json-schema
these are the current schema versions available.

I can add 1.27 / 1.28 if desired, and / or update to `v1.25.16` and `v1.26.13`.

Note that `.github/workflows/go-test.yaml` also has a matrix of Kube versions. If someone wants to provide a list of what versions are wanted for either / both, I can update them, but I'll probably default to not touching more things than necessary.

Comment (clamoriniere) [reply .github/workflows/ci.yaml:89]: If we can we would like to keep Kubernetes 1.16 in our tests. 
Unfortunately we still have some user with this version, but this version is the pre-version of several resources deprecation in the API server.

File: .github/workflows/ci.yaml
Comment (clamoriniere) [reply .github/workflows/ci.yaml:38]: should be ok

File: .github/workflows/ci.yaml
Comment (clamoriniere) [reply .github/workflows/ci.yaml:89]: If `kubeconform` is not available for 1.16, maybe we can keep kubeval for 1.16, and use `kubeconform` for the more ressent versions.  WDYT

File: charts/datadog/ci/kubeconform-values.yaml
Comment (clamoriniere) [inline charts/datadog/ci/kubeconform-values.yaml:39]: does this change is needed?

Comment (wyardley) [reply charts/datadog/ci/kubeconform-values.yaml:39]: See comment above and #1312 - short answer is yes if we want to have `-strict` and the intent is to allow multiple (different) values to be passed in to `nodeSelector`.

**Comment by wyardley** [comment]
#360 is technically already fixed, but I'm adding it in here as it will be extra-fixed once this is included.

File: charts/datadog/ci/kubeconform-values.yaml
Comment (clamoriniere) [reply charts/datadog/ci/kubeconform-values.yaml:39]: thanks for the explanation.

**Comment by clamoriniere** [comment]
Hi @wyardley 

thanks you for this contribution and your reactivity on comments. 
I made this PR on your branch https://github.com/wyardley/helm-charts/pull/1/ to add the missing change to make ou CI happy. If you can merge it. 

Then I think we will be able to merge the PR.

File: charts/datadog/Chart.yaml
Comment (clamoriniere) [inline charts/datadog/Chart.yaml:3]:
```suggestion
version: 3.53.4
```

File: charts/datadog/README.md
Comment (clamoriniere) [inline charts/datadog/README.md:3]:
```suggestion
![Version: 3.53.4](https://img.shields.io/badge/Version-3.53.4-informational?style=flat-square) ![AppVersion: 7](https://img.shields.io/badge/AppVersion-7-informational?style=flat-square)
```

File: .github/workflows/ci.yaml
Comment (wyardley) [inline .github/workflows/ci.yaml:85]: note: https://github.com/instrumenta/kubernetes-json-schema doesn't seem to have 1.16.15; 1.16.4 is the latest. Presumably this is Good Enough, but if someone can find the right schema version (I'm surprised `kubeval` worked with this) online in the right format, I can update the secondary schema location 🤷

**Comment by wyardley** [comment]
I rebased locally and slightly tweaked the wording, since there were conflicts from something merged after the previous version bump. Sorry for squashing your changes @clamoriniere

**Comment by wyardley** [comment]
Seems like it wanted additional bumps?
I bumped `charts/datadog-crds` and `charts/datadog-operator` too, in a separate commit, but let me know if you want me to revert it to how it was.

File: charts/datadog/Chart.yaml
Comment (clamoriniere) [inline charts/datadog/Chart.yaml:3]:
```suggestion
version: 3.54.1
```

File: charts/datadog/README.md
Comment (clamoriniere) [inline charts/datadog/README.md:3]:
```suggestion
![Version: 3.54.1](https://img.shields.io/badge/Version-3.54.1-informational?style=flat-square) ![AppVersion: 7](https://img.shields.io/badge/AppVersion-7-informational?style=flat-square)
```

File: charts/datadog/CHANGELOG.md
Comment (clamoriniere) [inline charts/datadog/CHANGELOG.md:3]:
```suggestion
## 3.54.1
```

File: .github/workflows/ci.yaml
Comment (clamoriniere) [inline .github/workflows/ci.yaml:118]: if seems that the kind node image 1.25.14 is not available 

but v1.25.16 is

File: .github/workflows/ci.yaml
Comment (clamoriniere) [inline .github/workflows/ci.yaml:119]: this version is not available but v1.26.13 it

```suggestion
          - v1.26.13
```

File: charts/datadog/ci/kubeconform-values.yaml
Comment (clamoriniere) [inline charts/datadog/ci/kubeconform-values.yaml:39]: I think the [CI is failing](https://github.com/DataDog/helm-charts/actions/runs/7872214797/job/21479319052?pr=1199#step:6:43) because we use this file in the CI Kind cluster. but the node doesn't have this nodeSelector. 

my suggestion for now is to rename this file: `charts/datadog/ci/kubeconform.yaml` so it will not be selected for the CI installation.

Comment (wyardley) [reply charts/datadog/ci/kubeconform-values.yaml:39]: I did that, but failed because I hadn't renamed it for each directory. But I do think there's some value to making sure to test / lint the values files, so since it looks like you've got a different / better fix in #1316 anyway, I reverted the change.

If you don't want it to be setting the disk type, I think we could probably set something arbitrary or harmless instead?

Comment (clamoriniere) [reply charts/datadog/ci/kubeconform-values.yaml:39]: Thanks @wyardley 

I'm still on it with another PR https://github.com/DataDog/helm-charts/pull/1317. Hopefully it will be the last one :)

**Review by clamoriniere — APPROVED** [review]
Hi @wyardley 
thanks again for all your work on this PR. Our CI is finally green ! 🥳 
We will be able to merge it

**Comment by wyardley** [comment]
Thanks all!

**Pull request merged**
